### PR TITLE
Querier: Reduce allocations on `QueryStreamResponse` unmarshal

### LINKED ISSUE #3265: Querier: High object allocation count from `QueryStreamResponse` unmarshal
## Problem

After obtaining a memory profile from one of the ruler dedicated queriers we've observed that approximately `33%` of total allocated objects correspond to timeseries deserialization when receiving stream responses from ingesters.

Screenshot 2022-10-20 at 13 23 31

Screenshot 2022-10-20 at 13 24 10

This particularly affects to ruler query pipelines, as rule expressions usually end up hitting ingesters when being resolved. 

## Proposal

Design some solution to reuse the largest possible portion of the timeseries objects when deserializing ingester stream responses, in order to reduce total generated garbage.

### PR DESCRIPTION
**Pull request by ortuman**
Signed-off-by: Miguel Ángel Ortuño <ortuman@gmail.com>

#### What this PR does

Use custom query stream response unmarshal code to allow reusing chunk slices when querying ingesters. 

To test the changes, new `BenchmarkQueryStreamResponseUnmarshal` has been added and executed with the following command: 

```sh
go test -bench=^BenchmarkQueryStreamResponseUnmarshal$ -run=X -cpu=8 -count=10 ./pkg/ingester/client
```

Below are the results for benchmark comparison between the previous and the new implementation: 

```
name                                                               old time/op    new time/op    delta
QueryStreamResponseUnmarshal/10_series,_10_chunks,_100_samples-8     51.6µs ± 1%    33.6µs ± 0%  -34.91%  (p=0.000 n=9+9)
QueryStreamResponseUnmarshal/10_series,_100_chunks,_100_samples-8    76.5µs ± 1%    41.6µs ± 0%  -45.61%  (p=0.000 n=10+9)
QueryStreamResponseUnmarshal/100_series,_1k_chunks,_1k_samples-8     4.43ms ± 0%    2.48ms ± 1%  -44.07%  (p=0.000 n=9+9)
QueryStreamResponseUnmarshal/100_series,_10k_chunks,_1k_samples-8    26.7ms ± 0%    11.5ms ± 1%  -56.68%  (p=0.000 n=9+10)
QueryStreamResponseUnmarshal/1k_series,_10k_chunks,_10k_samples-8     349ms ± 8%     271ms ± 2%  -22.17%  (p=0.000 n=9+10)

name                                                               old alloc/op   new alloc/op   delta
QueryStreamResponseUnmarshal/10_series,_10_chunks,_100_samples-8      143kB ± 0%      47kB ± 0%  -67.38%  (p=0.000 n=10+9)
QueryStreamResponseUnmarshal/10_series,_100_chunks,_100_samples-8     250kB ± 0%      47kB ± 0%  -81.37%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/100_series,_1k_chunks,_1k_samples-8     21.5MB ± 0%     5.1MB ± 0%  -76.05%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/100_series,_10k_chunks,_1k_samples-8     214MB ± 0%       7MB ± 1%  -96.56%  (p=0.000 n=10+9)
QueryStreamResponseUnmarshal/1k_series,_10k_chunks,_10k_samples-8    2.77GB ± 0%    1.21GB ± 0%  -56.46%  (p=0.000 n=10+8)

name                                                               old allocs/op  new allocs/op  delta
QueryStreamResponseUnmarshal/10_series,_10_chunks,_100_samples-8        221 ± 0%        93 ± 0%  -57.92%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/10_series,_100_chunks,_100_samples-8       251 ± 0%        93 ± 0%  -62.95%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/100_series,_1k_chunks,_1k_samples-8      3.22k ± 0%     1.24k ± 0%  -61.61%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/100_series,_10k_chunks,_1k_samples-8     3.92k ± 0%     1.25k ± 0%  -68.01%  (p=0.000 n=10+10)
QueryStreamResponseUnmarshal/1k_series,_10k_chunks,_10k_samples-8     46.0k ± 0%     26.3k ± 0%  -42.83%  (p=0.000 n=10+10)
```

#### Which issue(s) this PR fixes or relates to

Fixes #3265 

#### Checklist

- [X] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/distributor_queryable.go
Comment (56quarters) [inline pkg/querier/distributor_queryable.go:194]: I haven't take an in-depth look at this PR but I think it's worth confirming we make sure to call `Close()` on queriers everywhere. I seem to remember this wasn't always the case but we've gotten away with it because the existing ones don't leak anything.

File: pkg/ingester/client/wrapped.go
Comment (colega) [inline pkg/ingester/client/wrapped.go:174]: This seems to be autogenerated code, but it's unclear where it comes from or how to rebuild it again.

IMO having a code like this isn't worth it unless there's a HUGE performance improvement.

Comment (pstibrany) [reply pkg/ingester/client/wrapped.go:174]: > IMO having a code like this isn't worth it unless there's a HUGE performance improvement.

My first reaction was the same, but 1) we already have similar code (https://github.com/grafana/mimir/blob/062807abb982320d603fe7661fa6f2e1e581f7bc/pkg/mimirpb/timeseries.go#L131) and 2) benchmarks here don't look bad (although they need proper review too).

Comment (colega) [reply pkg/ingester/client/wrapped.go:174]: I agree that the benchmarks look good, then if we're keeping this code, we should document:
- What's the purpose of each of the new structs.
- How did we generate the unreadable Unmarshal() method (see the comment on the one @pstibrany linked)
- What are the expectations from the user of this entity (chunks have to be put back into the pool).

File: pkg/ingester/client/wrapped.go
Comment (colega) [inline pkg/ingester/client/wrapped.go:24]: What is the point of the wrapped response receiver? Why would I use it instead of the `Ingester_QueryStreamClient`?

File: pkg/querier/distributor_queryable.go
Comment (ortuman) [reply pkg/querier/distributor_queryable.go:195]: Found two more readers with a noop `Close()` method in our codebase. However those doesn't seem to be leaking anything, so I guess it's ok as they are. ([1](https://github.com/grafana/mimir/blob/main/pkg/querier/blocks_store_queryable.go#L432), [2](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/sharded_queryable.go#L146)).

File: pkg/distributor/query.go
Comment (ortuman) [reply pkg/distributor/query.go]: renamed @ 264d112a52876f4a2048fdcd3606efca28fb7f6e

**Review by colega — APPROVED** [review]

File: pkg/ingester/client/wrapped.go
Comment (ortuman) [reply pkg/ingester/client/wrapped.go:24]: added extra comment @ 264d112a52876f4a2048fdcd3606efca28fb7f6e

hope it's clearer now.

**Comment by ortuman** [comment]
FTR. I gave a shot to this new implementation in one of our dev environments (`cortex-dev-01`) and I'd say the difference is at least noticeable.

Below I've included screenshots of the profiles comparing the old and new implementations, both corresponding to `ruler-querier` component:

Before: 
before-ruler-querier

After:
after-ruler-querier

After comparison, we see a reduction of `16.66%` in total number of allocated objects when unmarshaling stream responses.

Worth adding, that memory allocation drop can also be noticed when looking at the GC pause rate.

Screenshot 2022-10-27 at 13 03 07

File: pkg/ingester/client/compat.go
Comment (ortuman) [reply pkg/ingester/client/compat.go]: Ahh, that's right. 

Changed @ 09064391afaba709f66d217ed4b881c661501f1d

File: pkg/ingester/client/wrapped.go
Comment (pracucci) [inline pkg/ingester/client/wrapped.go:421]: > len(qr.Chunkseries[i].Labels) > maxInPoolLabelsSliceSize

I don't fully understand what you're trying to do here. `maxInPoolLabelsSliceSize` is a very big number (16k) so we're not putting back in the pool slides bigger than 16k, but we can't have series with 16k labels. We validate input series and by default we limit to 30 labels per series.

Could you help me understand what we want to achieve here?

Comment (pracucci) [inline pkg/ingester/client/wrapped.go:421]: > len(qr.Chunkseries[i].Chunks) > maxInPoolChunksSliceSize

Should be `cap()` instead? I think all in all we care about the allocated capacity.

Comment (pracucci) [inline pkg/ingester/client/wrapped.go:421]: > maxInPoolChunksSliceSize

This value is 32K. We're saying at most we should reuse chunks slices up to 32K elements each. Chunks are per-series, so this mean we expect series from ingesters with more than 32K chunks. A chunk is full when it contains 120 samples, so 32K is about `32K * 120 samples = 3.8M samples`. Assuming a scraping interval of 10s, that's `3.8M * 10s = 439 days` of samples, but we only query the last 12h from ingesters.

Could you help me understand what we want to achieve here?

File: pkg/ingester/client/wrapped.go
Comment (pracucci) [inline pkg/ingester/client/wrapped.go:381]: Thinking loudly: the chunk structs in the reused `m.Chunks` are **not** reset, so they still contain the previous chunk data. I'm scared that if, for any reason, we receive a chunk with partial data in the wire format (e.g. due to future changes to the chunk proto) then the call to `Unmarshal()` will overwrite only the fields received over the wire.

What if we reset the chunk too in `ReuseQueryStreamResponse()`, keeping the chunk data `[]byte` but resetting its length to 0 (capacity is preserved)?

File: pkg/distributor/query.go
Comment (pracucci) [inline pkg/distributor/query.go:317]: Reading the code I couldn't figure out where the optimization comes from given we're making a copy of the `TimeSeriesChunk`. The optimization is given by the fact that `accumulateChunks()` (called previously) does a deduplication of identical chunks. Thanks to the RF=3, many chunks are perfectly identical and so the optimization is effective because we only copy `TimeSeriesChunk` after having deduplicated it.

Please update the comment to clearly explain how this optimization works.

File: pkg/distributor/query.go
Comment (pracucci) [inline pkg/distributor/query.go:330]: Have you considered moving it to a `defer()` (defined above) so that we reuse responses also when this function returns an error (e.g. limits have been reached)?

File: pkg/distributor/query.go
Comment (pracucci) [inline pkg/distributor/query.go:262]: I would rename the variables to make it clear it's unsafe to retain such memory.

```suggestion
			var unsafeResp ingester_client.WrappedQueryStreamResponse
```

Comment (pracucci) [reply pkg/distributor/query.go:262]: I would do the same for the response variable propagated through the chan in this function.

**Review by pracucci — COMMENTED** [review]
Good job! It took me a while to review it, but I can understand where the optimization comes from (see my comment about adding a comment to explain it clearly). I left few comments.

**Comment by pracucci** [comment]
Have you also considered adding a correctness test? I would call `QueryStream()` concurrently many times (with a cap to concurrency, e.g. 10 concurrent requests, for a total of 1K requests). Each call to `QueryStream()` should pass a different set of matchers. The mocked ingester should return a different result based on the label matchers (with variable number of series labels and chunks). Then each result of `QueryStream()` should be checked to make sure it matches the expected data.

File: pkg/distributor/query.go
Comment (ortuman) [reply pkg/distributor/query.go:330]: Didn't do that initially to avoid a race condition, as the reading goroutine could still be writing to the `responses`  slice when an error occurs. 

However, having a second look, I think we could simply hang on the `doneReading` channel at the begining of the `defer()` function to avoid this type of situation.

Changing it. 👍

File: pkg/ingester/client/wrapped.go
Comment (ortuman) [reply pkg/ingester/client/wrapped.go:421]: Intention here is trying to avoid sending big slices back to the pool, so if either labels or chunks backed arrays have a capacity bigger than a given threshold, we let it to be reclaimed by the GC.

Nevertheless, if we are fully confident that those stream responses are never going to grow beyond certain limit, I guess we could simply drop this guard. WDYT?

Comment (ortuman) [reply pkg/ingester/client/wrapped.go:421]: See: https://github.com/grafana/mimir/pull/3295#discussion_r1013893742

Comment (ortuman) [reply pkg/ingester/client/wrapped.go]: Discussed offline. We can safely drop this cap check. 👍

**Comment by ortuman** [comment]
@pracucci check new correctness test introduced at 3e572e0b1c60b0c8d34c722b110827da57245ceb

File: pkg/distributor/distributor_test.go
Comment (aknuds1) [inline pkg/distributor/distributor_test.go:4550]: For my own understanding, what's the benefit of running the queries concurrently as part of the test?

File: pkg/ingester/client/compat.go
Comment (aknuds1) [inline pkg/ingester/client/compat.go:266]: It wouldn't be easier/equivalent (plus you won't have to state each chunk field explicitly) here to just copy `ts.Chunks`? Let me know if I didn't think of something :)

```suggestion
	copy(chunks, ts.Chunks)
	for c, j := range ts.Chunks {
		chunks[j].Data = make([]byte, len(c.Data))
		copy(chunks[j].Data, c.Data)
	}
```

**Review by aknuds1 — COMMENTED** [review]
Left some comments.

File: pkg/ingester/client/wrapped.go
Comment (aknuds1) [reply pkg/ingester/client/wrapped.go:196]: Agree with @colega - I'd also like to see documentation of the origin of auto-generated code. I could only deduce this is possibly (partially?) auto-generated. E.g. which portions of this code is auto-generated, and which are not?

File: pkg/distributor/distributor_test.go
Comment (colega) [reply pkg/distributor/distributor_test.go:4520]: Then I'd suggest
```suggestion
		now := time.Now().Add(-2 * time.Hour)
```

As that's human-readable (at least in English, Spanish and Russian, IDK about other languages).

Comment (aknuds1) [reply pkg/distributor/distributor_test.go:4520]: Agree with @colega.

**Comment by replay** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/distributor/distributor_test.go
Comment (ortuman) [reply pkg/distributor/distributor_test.go:4550]: The `QueryStream` method invokes `queryIngesterStream` under the hood, which in turn performs the ingesters's [stream response reuse](https://github.com/grafana/mimir/blob/1f219b2c765d5b89e8a40138590b5c84ed869494/pkg/distributor/query.go#L196-L199).

So the purpose of running this concurrently is to simulate a more realistic scenario, and to validate that nothing messes up when querying ingesters.

Comment (aknuds1) [reply pkg/distributor/distributor_test.go:4550]: @ortuman thanks for clarifying! That's pretty much what I thought. Maybe you could also note this rationale in a comment (below the existing one)?

File: pkg/ingester/client/compat.go
Comment (ortuman) [reply pkg/ingester/client/compat.go:266]: Problem with that is that we wouldn't be creating a deep copy of the original `TimeSeriesChunks` object, since `Data` backed buffer would be shared between source and destination chunks, which can cause some unexpected behaviors when reusing those from pool.

File: pkg/distributor/distributor_test.go
Comment (ortuman) [reply pkg/distributor/distributor_test.go:4550]: 'course! I'll include it. 👍

Comment (ortuman) [reply pkg/distributor/distributor_test.go]: done @ 74f97b01b9abb71759f5343487e411da8202c342

File: pkg/ingester/client/compat.go
Comment (ortuman) [reply pkg/ingester/client/compat.go:266]: Oh, sorry. I misread it completely. 🤦‍♂️  

And yes, I agree, that's equivalent and more concise. 👍

Comment (aknuds1) [reply pkg/ingester/client/compat.go]: Yay :D

**Review by charleskorn — COMMENTED** [review]
Late to the party here, but I'm interested in this in the context of https://github.com/grafana/mimir/pull/4886 and the querier optimisations we're doing.

This looks good to me, I'd like to see two things before taking this further:

* a rebase on `main` (particularly to get the changes in https://github.com/grafana/mimir/pull/5016)
* a benchmark comparing `queryIngesterStream` before and after this change (as `BenchmarkQueryStreamResponseUnmarshal` doesn't include the cost of later copying chunks if we do want to retain them) to see if the additional maintenance burden is worth the performance gain

We'll also need to do some work to add the new `QueryStreamResponse` fields added in https://github.com/grafana/mimir/pull/4886 once that is merged.

**Comment by pracucci** [comment]
My main concern around any optimization like this, is the copying we have to do for the protobuf generated code (`pkg/ingester/client/wrapped.go`) which makes it hard to maintain. I know we're already doing it in other places, and I have concerns too. If benchmarks are good, I will **not** block on this, but longer term we should have a better strategy.

**Comment by ortuman** [comment]
Following @charleskorn suggestion, I've added a new benchmark named `BenchmarkDistributor_QueryStream` to this PR in order to test the complete execution flow for the `QueryStream` method. After rebasing with the main branch and adapting the code to recent changes, I've observed that the performance of this branch is lower compared to the current implementation.

Results:
```
Distributor_QueryStream/10_metrics,_100_samples-8     162µs ± 3%     219µs ± 1%  +35.41%  (p=0.000 n=10+10)
Distributor_QueryStream/100_series,_1k_samples-8     6.37ms ± 1%   10.64ms ± 0%  +66.90%  (p=0.000 n=10+10)
Distributor_QueryStream/1k_series,_1k_samples-8       198ms ± 1%     301ms ± 6%  +51.85%  (p=0.000 n=9+9)

name                                               old alloc/op   new alloc/op   delta
Distributor_QueryStream/10_metrics,_100_samples-8     304kB ± 0%     398kB ± 0%  +30.73%  (p=0.000 n=10+10)
Distributor_QueryStream/100_series,_1k_samples-8     15.4MB ± 0%    20.3MB ± 0%  +31.32%  (p=0.000 n=10+10)
Distributor_QueryStream/1k_series,_1k_samples-8       320MB ± 0%     443MB ± 0%  +38.48%  (p=0.000 n=10+8)

name                                               old allocs/op  new allocs/op  delta
Distributor_QueryStream/10_metrics,_100_samples-8     4.19k ± 0%     5.42k ± 0%  +29.47%  (p=0.000 n=10+7)
Distributor_QueryStream/100_series,_1k_samples-8       199k ± 0%      260k ± 0%  +30.53%  (p=0.000 n=10+10)
Distributor_QueryStream/1k_series,_1k_samples-8       3.99M ± 0%     5.58M ± 0%  +39.83%  (p=0.000 n=10+8)
```

My assumption here is that the performance gained when unmarshaling the timeseries chunks is subsequently lost due to the need to make a deep copy before returning the result. I understand that the recent changes have outperformed the advantages introduced in this branch, so if you think it's appropriate, you can close it. 

And by the way, great job. 😉

File: pkg/distributor/query_test.go
Comment (charleskorn) [inline pkg/distributor/query_test.go:679]: I've found calls to `require` to be surprisingly expensive (it has to do some locking to deal with multi-threaded tests) - have you checked a CPU profile to see if the cost of these calls is significant here (and therefore skewing the benchmark)?

Comment (pracucci) [reply pkg/distributor/query_test.go:679]: The TL;DR is: do not use any `require` or `assert` in any benchmarked code. If you want to check for an error, just do `if err != nil { b.Fail(...) }`.

**Pull request closed without merging**
